### DISCUSSION TITLE
Overriding zoom-button menu

### DISCUSSION THREAD

**Discussion by gdinh** [Hammerspoon/hammerspoon#3354] [Q&A]
I'd like to override the dialog that pops up when you hover over the zoom button on windows:

<img width="280" alt="image" src="https://user-images.githubusercontent.com/34658064/210283696-121c00e4-047a-4336-a556-f8f77dc5bcf3.png">

For instance, Moom manages to replace the dialog with one of its own:

<img width="280" alt="image" src="https://i.imgur.com/ZZAmpYm.jpg">

In particular, I'm wondering:

- how to determine when the mouse is hovering over the zoom button (seems like something to do with axuielement.observer, but I'm not sure what to tell it to watch for - also, would require watching for applications opening/closing in order to set up a new watcher for each one - is there a performance penalty for doing this?)
- how to suppress display of the system menu when the spoon is active
- whether it's possible to override the color of the zoom button to serve as a status indicator.

Thank you!

**Comment by Rhys-T** [comment]
Disabling the builtin menu is the easy part - there's a hidden setting to do that:
```shell
$ defaults write -g NSZoomButtonShowMenu -bool no
$ # to bring it back:
$ defaults delete -g NSZoomButtonShowMenu
```
Note that apps only seem to read this when they launch, so you'll probably want to log out and back in (or restart) to make sure everything picks up the new setting. (Note: there's an alternate approach below, if you want to only disable the standard menu while Hammerspoon is running.)

The other parts are going to be somewhat trickier, unfortunately. I tried attaching an `hs.axuielement.observer` to every notification in [the `notifications` table][notifs] for the TextEdit app, its window, _and_ that window's zoom/fullscreen button, but none of them seemed to give any indication of when the button was being hovered over - or even when the builtin menu appeared, so I could immediately close and replace it. It's possible that it sends some other notification that's not in that list, but I don't know of any good way to find it if so. (Also, I'm still on Catalina, and it's possible that one of the standard notifications _is_ sent on newer systems.)

Failing that, the only approach I know of would be to use an [`hs.eventtap`][eventtap] to watch for `mouseMoved` events, and use [`systemElementAtPosition`][syselat] to see whether what you're pointing at has an `AXSubrole` of `AXFullScreenButton` or `AXZoomButton`[^subroles]… all of which could obviously add a fair amount of overhead. If you do decide to do that, you'll probably want to lower the timeout while you check for the button - some apps (particularly ones built on certain cross-platform frameworks) tend to take a relatively long time to respond to `systemElementAtPosition`. You can also use the event tap to prevent the builtin menu instead of using the setting above (so that the normal menu comes back if Hammerspoon isn't running), by modifying the event's location to not be over the button.

<details>
<summary>Example of watching for zoom button hover using <code>hs.eventtap</code></summary>

```lua
local sysEl = hs.axuielement.systemWideElement()

-- temporarily reduce the timeout for the Accessibility API
-- and make sure it gets set back even if there's an error
local fastAXResetter = setmetatable({}, {
	__close = function()
		sysEl:setTimeout(0) -- set timeout back to default
	end
})
local function useFastAX()
	sysEl:setTimeout(.0025)
	return fastAXResetter
end

local zoomButtonSubroles = {
	AXFullScreenButton = true,
	AXZoomButton = true,
}

local wasZoomButton = false
local zoomButtonOffLeftEdgeX -- OPTIONAL
zoomButtonHoverTap = hs.eventtap.new({hs.eventtap.event.types.mouseMoved}, function(event)
	local _ <close> = useFastAX() -- automatically reverts at the end of this function
	local location = event:location()
	local pointedElement = sysEl:elementAtPosition(location)
	local isZoomButton = pointedElement and zoomButtonSubroles[pointedElement.AXSubrole]
	if isZoomButton and not wasZoomButton then
		wasZoomButton = true
		print('enter zoom button')
		-- start timer here
	elseif wasZoomButton and not isZoomButton then
		wasZoomButton = false
		print('exit zoom button')
		zoomButtonOffLeftEdgeX = nil -- OPTIONAL
		-- cancel timer here
	end
	
	-- OPTIONAL:
	-- pretend the cursor is just to the left of the zoom/fullscreen button
	-- keeps the menu from appearing, but lets the button symbols appear normally
	if isZoomButton then
		if not zoomButtonOffLeftEdgeX then
			zoomButtonOffLeftEdgeX = pointedElement.AXPosition.x - 1
		end
		event:location({x = zoomButtonOffLeftEdgeX, y = location.y})
	end
end):start()
```

</details>

Changing the button's color could probably be accomplished (sort of) by tracking its position with an [`hs.window.filter`][winfilt], and drawing a recolored version of the button over the real one using an [`hs.canvas`][canvas].  An event tap could watch for the cursor to approach the buttons, and handle showing/hiding the symbol on the fake button. Just be aware that the canvas won't track very smoothly while you're dragging the window. (I ran into this while playing with the idea of adding a fourth button to certain windows.) It'll also be somewhat trickier if you want it to apply to more than just the focused window.

[^subroles]: Ever since the zoom and fullscreen buttons were merged in Yosemite, the subrole has been `AXFullScreenButton`… but when a window is already fullscreen, the green button to _exit_ fullscreen is an `AXZoomButton`. Go figure.

[notifs]: https://www.hammerspoon.org/docs/hs.axuielement.observer.html#notifications
[eventtap]: https://www.hammerspoon.org/docs/hs.eventtap.html
[syselat]: https://www.hammerspoon.org/docs/hs.axuielement.html#systemElementAtPosition
[winfilt]: https://www.hammerspoon.org/docs/hs.window.filter.html
[canvas]: https://www.hammerspoon.org/docs/hs.canvas.html
